pkg/e2e/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,371 @@
+				glog.Errorf("Machine %q does not have node reference set", machine.Name)
+				return false, nil
+			}
+			node := corev1.Node{}
+			if err := client.Get(context.TODO(), types.NamespacedName{Name: machine.Status.NodeRef.Name}, &node); err != nil {
+				glog.Errorf("error querying api for node object: %v, retrying...", err)
+				return false, nil
+			}
+
+			labelNode := false
+			if node.Labels == nil {
+				labelNode = true
+			} else if _, exists := node.Labels[nodeTestLabel]; !exists {
+				labelNode = true
+			}
+
+			if labelNode {
+				nodeCopy := node.DeepCopy()
+				if nodeCopy.Labels == nil {
+					nodeCopy.Labels = make(map[string]string)
+				}
+				nodeCopy.Labels[nodeTestLabel] = ""
+				if err := client.Update(context.TODO(), nodeCopy); err != nil {
+					glog.Errorf("error updating api for node object: %v, retrying...", err)
+					return false, nil
+				}
+				glog.Infof("Labeling node %q with %q label", nodeCopy.Name, nodeTestLabel)
+			}
+		}
+		return true, nil
+	})
+}
+
+var _ = g.Describe("[Feature:Machines] Autoscaler should", func() {
+	defer g.GinkgoRecover()
+
+	g.It("scale out", func() {
+		var err error
+		client, err := e2e.LoadClient()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		glog.Info("Get one machineSet")
+		machineSetList := mapiv1beta1.MachineSetList{}
+		err = wait.PollImmediate(1*time.Second, waitMedium, func() (bool, error) {
+			if err := client.List(context.TODO(), runtimeclient.InNamespace(e2e.TestContext.MachineApiNamespace), &machineSetList); err != nil {
+				glog.Errorf("error querying api for nodeList object: %v, retrying...", err)
+				return false, err
+			}
+			return len(machineSetList.Items) > 0, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// When we add support for machineDeployments on the installer, cluster-autoscaler and cluster-autoscaler-operator
+		// we need to test against deployments instead so we skip this test.
+		targetMachineSet := machineSetList.Items[0]
+		if ownerReferences := targetMachineSet.GetOwnerReferences(); len(ownerReferences) > 0 {
+			// glog.Infof("MachineSet %s is owned by a machineDeployment. Please run tests against machineDeployment instead", targetMachineSet.Name)
+			g.Skip(fmt.Sprintf("MachineSet %s is owned by a machineDeployment. Please run tests against machineDeployment instead", targetMachineSet.Name))
+		}
+
+		glog.Infof("Create ClusterAutoscaler and MachineAutoscaler objects. Targeting machineSet %s", targetMachineSet.Name)
+		initialNumberOfReplicas := targetMachineSet.Spec.Replicas
+		clusterAutoscaler := caov1alpha1.ClusterAutoscaler{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "default",
+				Namespace: e2e.TestContext.MachineApiNamespace,
+			},
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ClusterAutoscaler",
+				APIVersion: "autoscaling.openshift.io/v1alpha1",
+			},
+			Spec: caov1alpha1.ClusterAutoscalerSpec{
+				ScaleDown: &caov1alpha1.ScaleDownConfig{
+					Enabled:           true,
```

Comment:
the autoscaler is expecting pointers to strings here, so this will break

Reply:
```go
type ScaleDownConfig struct {
	Enabled           bool   `json:"enabled"`
	DelayAfterAdd     string `json:"delayAfterAdd"`
	DelayAfterDelete  string `json:"delayAfterDelete"`
	DelayAfterFailure string `json:"delayAfterFailure"`
	UnneededTime      string `json:"unneededTime,omitempty"`
}
```

Reply:
Not sure where you got that from. Please see https://github.com/openshift/cluster-autoscaler-operator/blob/master/pkg/apis/autoscaling/v1alpha1/clusterautoscaler_types.go#L67

Reply:
Changed by this commit [1]. The `vendor` dir has old version of CAO code.

[1] https://github.com/openshift/cluster-autoscaler-operator/commit/2a05e7b63d9eedcfb084262bbdd138c57a4723da#diff-10d8058f9e03d0da119d441627f8ac4d

Reply:
yes, so it needs a newer version so we can run the tests against cao

Reply:
Done